### PR TITLE
feat: support rego in custom rules

### PR DESCRIPTION
**Pull request by TzlilSwimmer123**

### CODE REVIEW

File: pkg/jsonSchemaValidator/extensions/customKeyRegoRule.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
+// This file defines a custom key to implement the logic for the rule:
```
Comment (hadar-co) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoRule.go]: @TzlilSwimmer123 i think this file should be named "customKeyRegoDefinition" like the others whose name is the same name as the key

File: go.mod [hunk 2]
```diff
@@ -13,46 +13,53 @@ require (
...
 	gopkg.in/yaml.v2 v2.4.0
 	gopkg.in/yaml.v3 v3.0.1
 	k8s.io/apimachinery v0.23.5
 	sigs.k8s.io/yaml v1.2.0
 )
 
+require github.com/open-policy-agent/opa v0.49.2
+
 require (
+	github.com/OneOfOne/xxhash v1.2.8 // indirect
```
Comment (shmu3l) [inline go.mod:35]: are those extra packages are necessary? `xxhash`

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+			}
+			return nil
+		} else {
+			return ctx.Error(RegoDefinitionCustomKey, "violation needs to return a boolean")
+		}
+	}
+}
+
+func getPackageFromRegoCode(regoCode string) string {
+	const PACKAGE = "package"
```
Comment (shmu3l) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:98]: if the word "package" dose not exist in `regoCode`, it can be a panic error?

image

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const RegoDefinitionCustomKey = "regoDefinition"
+
+var regoCodeToEval = RegoDefinition{}
```
Comment (shmu3l) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: im not sure about the global var that all function can modify it

same case but a lil bit different, maybe it can explain better: 
https://gobyexample.com/mutexes

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	index := strings.Index(regoCode, PACKAGE)
+	// get next single word after "package"
+	packageStr := strings.Fields(regoCode[index:])
+	return packageStr[1]
+}
+
+func retrieveRegoFromSchema(regoCode string) *rego.Rego {
+	mainRegoPackage = getPackageFromRegoCode(regoCode)
+
+	var regoObjectParts []func(r *rego.Rego)
```
Comment (shmu3l) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:118]: `retrieveRegoFromSchema` - It is assumed that the package is in the libs field of the `regoCode`, but if it is not, would this code still work? It might be better to verify that the `package` key is in the `lib` as well.

same logic as `getPackageFromRegoCode`

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 3 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:98]: I don't think that panicing here will be good because imagine we have a policy with 6 different rules, 5 are ok and 1 is not, instead of evaluating the rule we will panic the policyCheck, feels like a bad experience

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 4 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: Basically I need some way to preserve the rego code I get in the Compile function. Basically only 1 place in the code sets and no one updated it. What is the concern?

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 5 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:118]: Basically the rego code will not compile and in this situation we fail the rule. package `something` is a must

File: go.mod (hunk 2 above)
Comment (royhadad) [reply go.mod:35]: those seem to be indirect dependencies

Comment (TzlilSwimmer123) [reply go.mod:35]: I run `go mod tidy` and it didn't change

File: cmd/publish/main.go [hunk 6]
```diff
@@ -105,11 +116,46 @@ type MessagesContext struct {
...
+		if !valid {
+			return nil, fmt.Errorf("failed to validate rego custom rules")
+		}
+	}
+
 	return ctx.PublishCliClient.PublishPolicies(policiesConfiguration, localConfigContent.Token)
 }
+
+func validateIfRegoRules(ctx *PublishCommandContext, customRules []interface{}) (bool, error) {
+	for _, regoRule := range customRules {
```
Comment (royhadad) [inline cmd/publish/main.go]: at this stage we don't know yet if it's a regoRule
```suggestion
	for _, cusotmRule := range customRules {
```

File: cmd/publish/regoDefinitionSchema.yaml [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+      regoDefinition:
+        type: object
+        required:
+          - code
+        properties:
+          libs:
+            type: array
+          code:
+            type: string
+            pattern: \bviolation\b
```
Comment (royhadad) [inline cmd/publish/regoDefinitionSchema.yaml]: can we validate that the rule starts with `package xxx` as well?

File: cmd/publish/regoDefinitionSchema.yaml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+definitions:
+  regoPattern:
+    properties:
+      regoDefinition:
+        type: object
+        required:
+          - code
+        properties:
+          libs:
+            type: array
```
Comment (royhadad) [inline cmd/publish/regoDefinitionSchema.yaml:10]: validate the item type as well to be a string

File: cmd/publish/regoDefinitionSchema.yaml [hunk 9]
```diff
@@ -0,0 +1,19 @@
+definitions:
```
Comment (royhadad) [inline cmd/publish/regoDefinitionSchema.yaml:1]: I was thinking about extending the file `policiesSchema.json` with the rego definition, have you considered it?

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const RegoDefinitionCustomKey = "regoDefinition"
+
+var regoCodeToEval = RegoDefinition{}
+var mainRegoPackage = ""
+var mainModuleFileName = "main.rego"
+var regoFunctionEntryPoint = "violation"
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]:
```suggestion
const regoFunctionEntryPoint = "violation"
```

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const RegoDefinitionCustomKey = "regoDefinition"
+
+var regoCodeToEval = RegoDefinition{}
+var mainRegoPackage = ""
+var mainModuleFileName = "main.rego"
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]:
```suggestion
const mainModuleFileName = "main.rego"
```

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+	"strings"
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/santhosh-tekuri/jsonschema/v5"
+)
+
+const RegoDefinitionCustomKey = "regoDefinition"
+
+var regoCodeToEval = RegoDefinition{}
+var mainRegoPackage = ""
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: This variable is only used in 1 function, it can definitely become a private variable of that function, instead of being exposed to (and potentially mutated by) the entire module

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 4 above)
Comment (royhadad) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: I share @shmu3l's concern, this code smells funny. 
what's wrong with compiling the rego code in the validation function?
or maybe there is a way to pass a value from the `Compile` function to the `Validate` function?

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+	// Execute the prepared query.
+	rs, err := query.Eval(regoCtx, rego.EvalInput(dataValue))
+
+	if err != nil {
+		return ctx.Error(RegoDefinitionCustomKey, "failed to evaluate rego due to %s", err.Error())
+	}
+
+	if len(rs) != 1 || len(rs[0].Expressions) != 1 {
+		return ctx.Error(RegoDefinitionCustomKey, "failed to evaluate rego, unexpected results")
+	} else {
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: a bit easier on the eyes
```suggestion
```

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+		resultsValue := (rs[0].Expressions[0].Value).([]interface{})
+		value, ok := resultsValue[0].(bool)
+		if ok {
+			if value {
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: what is this `value`? what does it represent? maybe a different name?

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+	}
+}
+
+func getPackageFromRegoCode(regoCode string) string {
+	const PACKAGE = "package"
+	// find the index of string "package"
+	index := strings.Index(regoCode, PACKAGE)
+	// get next single word after "package"
+	packageStr := strings.Fields(regoCode[index:])
+	return packageStr[1]
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: why 1 and not 0?

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	return nil, nil
+}
+
+type RegoDefinition struct {
+	Libs []string `json:"libs"`
+	Code string   `json:"code"`
+}
+
+func (s CustomKeyRegoDefinitionSchema) Validate(ctx jsonschema.ValidationContext, dataValue interface{}) error {
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: what happens when this `Validate` function throws a panic?
maybe add an error handler as a fallback?
```
defer func() {
		if panicErr := recover(); panicErr != nil {
// do something                 
}
	}()
```

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 17]
```diff
@@ -0,0 +1,116 @@
+// This file defines a custom key to implement the logic for rego rule:
+
+package jsonSchemaValidator
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:3]: ***global comment***
Can we add some test coverage for this feature? (publish, test, --policy-config)
There are a lot of edge cases and I can't really know how the system will behave in each case

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 15 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: in zero index it's `package`

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 4 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: so I refactored it, I didn't see we do get it inside `Validate`

File: cmd/publish/main.go [hunk 18]
```diff
@@ -105,11 +116,46 @@ type MessagesContext struct {
...
+
 	return ctx.PublishCliClient.PublishPolicies(policiesConfiguration, localConfigContent.Token)
 }
+
+func validateIfRegoRules(ctx *PublishCommandContext, customRules []interface{}) (bool, error) {
+	for _, customRule := range customRules {
+		b, _ := json.Marshal(customRule)
+		regoRuleStr := string(b)
+
+		isRegoRule := strings.Contains(regoRuleStr, jsonSchemaValidator.RegoDefinitionCustomKey)
```
Comment (royhadad) [inline cmd/publish/main.go]: I think this `if` can be removed, to make it simpler. Simply put - we'll validate every rule against the regoDefinitionSchema. if the key is missing - the rule will pass.
then the function can be renamed to `validateRegoDefinitions`

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+func (customKeyRegoDefinitionSchema CustomKeyRegoDefinitionSchema) Validate(ctx jsonschema.ValidationContext, dataValue interface{}) error {
+	regoDefinitionSchema, err := convertCustomKeyRegoDefinitionSchemaToRegoDefinitionSchema(customKeyRegoDefinitionSchema)
+	if err != nil {
+		return ctx.Error(RegoDefinitionCustomKey, err.Error())
+	}
+
+	regoCtx := context.Background()
+
+	r, err := retrieveRegoFromSchema(regoDefinitionSchema)
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go]: please rename this `r`

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+const RegoDefinitionCustomKey = "regoDefinition"
+
+type CustomKeyRegoDefinitionCompiler struct{}
+
+type CustomKeyRegoDefinitionSchema map[string]interface{}
+
+var CustomKeyRegoRule = jsonschema.MustCompileString("customKeyRegoDefinition.json", `{
+	"properties" : {
+		"regoDefinition": {
+			"type": "object"
```
Comment (royhadad) [inline pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:24]: Have you considered putting the `regoDefinitionSchema.yaml` here?

**Review by royhadad — DISMISSED** [review]

File: pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go (hunk 20 above)
Comment (TzlilSwimmer123) [reply pkg/jsonSchemaValidator/extensions/customKeyRegoDefinition.go:24]: I think that atm it is connected only to publish command so I feel good to leave it there

**Review by royhadad — APPROVED** [review]

**Review by hadar-co — APPROVED** [review]
